### ISSUE TITLE
SPI issues

### ISSUE THREAD

**Issue by RudolphRiedel** [board707/w80x_arduino#25]
I am just playing with a W806 board and SPI.
I intend to look deeper but this is what I found now.

I have this little sketch:
#include <Arduino.h>
#include <SPI.h>

#define EVE_CS PB10
#define EVE_PDN PB11

void setup()
{
  pinMode(EVE_CS, OUTPUT);
  digitalWrite(EVE_CS, HIGH);
  pinMode(EVE_PDN, OUTPUT);
  digitalWrite(EVE_PDN, LOW);

  SPI.begin(); /* sets up the SPI to run in Mode 0 and 1 MHz */
  SPI.beginTransaction(SPISettings(8000000, MSBFIRST, SPI_MODE0));
}

void loop()
{
  delay(2);

  digitalWrite(EVE_CS, LOW);
  SPI.transfer(0x55);
  SPI.transfer(0xaa);
  SPI.transfer(0x55);
  digitalWrite(EVE_CS, HIGH);

  digitalWrite(EVE_CS, LOW);
  uint32_t buffer = 0x55aa55aa;
  SPI.transfer(&buffer, 3);
  digitalWrite(EVE_CS, HIGH);
}

And this is what I get in the logic analyzer:
![grafik](https://github.com/board707/w80x_arduino/assets/31180093/41af020b-a59a-4724-93ec-d67d683cab80)

So well, on a quick glance everything is working, there are a couple of issues though.
- the frequency of that SPI is not 8 MHz, it is 1MHz.
- the pause between the single byte transfers is over 7µs
- MOSI behaves a bit odd, like the default seems to be low but sometimes it goes high

![grafik](https://github.com/board707/w80x_arduino/assets/31180093/d112d867-af47-464c-b5a7-de92e17cc62f)
![grafik](https://github.com/board707/w80x_arduino/assets/31180093/8693aa61-61bc-4dea-9087-b3395726a088)
![grafik](https://github.com/board707/w80x_arduino/assets/31180093/75091480-c3a0-4397-9831-8f8c66c1812d)

I'll have a look to see what can be done about this.

**Comment by AnatolSher** [comment]
Hello
![image](https://github.com/board707/w80x_arduino/assets/121094544/483afd58-fe1e-4049-adb0-97cdeb323d79)
Default SS pin
![image](https://github.com/board707/w80x_arduino/assets/121094544/51baaf81-c732-4726-92aa-f1f4db89277a)
Doc - https://github.com/board707/w80x_arduino/blob/hal-v0.6.0/doc/SPI.md

**digitalWrite(EVE_CS, HIGH);
digitalWrite(EVE_CS, LOW);**
It takes a lot of time

When we select 20MHz for SPI, it means base peripheral clock frequency is 40mHz/prescaler
![image](https://github.com/board707/w80x_arduino/assets/121094544/3c9216d5-d595-4438-b55e-ee5c84f8919f)
As a result, we have a max spi clock frequency of 20 Mhz

Possible values ​​for spi clock are
![image](https://github.com/board707/w80x_arduino/assets/121094544/404f668c-2012-43a8-bd1c-b5d04a4fafb4)

If you need more see workaround in SDIO_TFT_Test folder

**Comment by RudolphRiedel** [comment]
Thanks for pointing out that SPI.md

Yes, I just worked thru SPI.cpp, reformatted it to a more consistence style and translated the few russian comments.
And while it is still working I noticed the limitations and switched to 10MHz which is working.

The pauses got a bit shorter with the increased SPI clock. 

![grafik](https://github.com/board707/w80x_arduino/assets/31180093/48f002ef-3fd0-4066-88a0-cb3b1949d759)

I will need a software controlled CS, so I just continue to use it.

**Comment by board707** [comment]
> I noticed the limitations and switched to 10MHz which is working.

Yes, the package probably lacks clear style and detailed documentation.
However, as far as I understand, the problem is generally solved?

**Comment by RudolphRiedel** [comment]
Nah, it's still rather slow, especially for a 240MHz core.
But that may be something to fix here:
https://github.com/IOsetting/wm-sdk-w806/blob/main/platform/drivers/wm_spi.c
Well, or with register based code in SPI.cpp.

I'll have a look, but not right away.

**Comment by board707** [comment]
> But that may be something to fix here:
> https://github.com/IOsetting/wm-sdk-w806/blob/main/platform/drivers/wm_spi.c

Our code based on the exact this file.
But anyway, it will be useful if you look on it again
